File: starrysky/src/main/java/com/lzx/starrysky/delayaction/PlayValidManager.java
package com.lzx.starrysky.delayaction;

import com.lzx.starrysky.StarrySky;
import com.lzx.starrysky.provider.SongInfo;
import com.lzx.starrysky.utils.StarrySkyUtils;

import java.util.ArrayList;
import java.util.List;

public class PlayValidManager {

    private static PlayValidManager playValidManager;
    private List<Valid> mValidQueue = new ArrayList<>();
    private Action action;
    private int validIndex = 0;
    private boolean doCallAfterAction = true;

    public static PlayValidManager get() {
        if (playValidManager == null) {
            synchronized (StarrySky.class) {
                if (playValidManager == null) {
                    playValidManager = new PlayValidManager();
                }
            }
        }
        return playValidManager;
    }

    public void addValid(Valid valid) {
        if (!mValidQueue.contains(valid)) {
            mValidQueue.add(valid);
        }
    }

    public PlayValidManager setAction(Action action) {
        if (action != null) {
            this.action = action;
        }
        return this;
    }

    public List<Valid> getValidQueue() {
        return mValidQueue;
    }

    public void doCall(SongInfo songInfo) {
        StarrySkyUtils.log("doCall#validIndex = " + validIndex);
        //执行验证
        if (validIndex < mValidQueue.size()) {
            Valid valid = mValidQueue.get(0);
            if (valid != null) {
                valid.doValid(songInfo, new Valid.ValidCallback() {
                    @Override
                    public void finishValid() {
                        validIndex++;
                        StarrySkyUtils.log("doCall#  validIndex++ ");
                        doCall(songInfo);
                    }

                    @Override
                    public void doActionDirect() {
                        StarrySkyUtils.log("直接执行 Action");
                        doAction(songInfo); //直接执行
                    }
                });
            }
        } else {
            //执行action
            doAction(songInfo, true);
        }
    }

    public void doAction(SongInfo songInfo) {
        doAction(songInfo, false);
    }

    private void doAction(SongInfo songInfo, boolean doCallAfterAction) {
        if (action != null) {
            this.doCallAfterAction = doCallAfterAction;
            action.call(songInfo);
            validIndex = 0;
            StarrySkyUtils.log("doAction#validIndex = " + validIndex);
        }
    }

    public void doCall(String mediaId) {
        SongInfo songInfo = StarrySky.get().getMediaQueueProvider().getSongInfo(mediaId);
        doCall(songInfo);
    }

    public void resetValidIndex() {
        validIndex = 0;
    }

    public void clear() {
        validIndex = 0;
        mValidQueue.clear();
    }
}

File: starrysky/src/main/java/com/lzx/starrysky/delayaction/PlayValidManager.kt
package com.lzx.starrysky.delayaction

import com.lzx.starrysky.StarrySky
import com.lzx.starrysky.provider.SongInfo
import com.lzx.starrysky.utils.StarrySkyUtils

class PlayValidManager {

    private val mValidQueue = mutableListOf<Valid>()
    private var action: Action? = null
    private var validIndex = 0
    private var doCallAfterAction = true

    companion object {
        @Volatile
        private var instance: PlayValidManager? = null

        fun get(): PlayValidManager {
            return instance ?: synchronized(this) {
                instance ?: PlayValidManager().also { instance = it }
            }
        }
    }

    fun addValid(valid: Valid) {
        if (!mValidQueue.contains(valid)) {
            mValidQueue.add(valid)
        }
    }

    fun setAction(action: Action?): PlayValidManager {
        if (action != null) {
            this.action = action
        }
        return this
    }

    fun getValidQueue(): List<Valid> {
        return mValidQueue
    }

    fun doCall(songInfo: SongInfo?) {
        StarrySkyUtils.log("doCall#validIndex = $validIndex")
        //执行验证
        if (validIndex < mValidQueue.size) {
            val valid = mValidQueue[0]
            valid.doValid(songInfo, object : Valid.ValidCallback {
                override fun finishValid() {
                    validIndex++
                    StarrySkyUtils.log("doCall#  validIndex++ ")
                    doCall(songInfo)
                }

                override fun doActionDirect() {
                    StarrySkyUtils.log("直接执行 Action")
                    doAction(songInfo) //直接执行
                }
            })
        } else {
            //执行action
            doAction(songInfo, true)
        }
    }

    fun doAction(songInfo: SongInfo?) {
        doAction(songInfo, false)
    }

    private fun doAction(songInfo: SongInfo?, doCallAfterAction: Boolean) {
        if (action != null) {
            this.doCallAfterAction = doCallAfterAction
            action?.call(songInfo)
            validIndex = 0
            StarrySkyUtils.log("doAction#validIndex = $validIndex")
        }
    }

    fun doCall(mediaId: String?) {
        val songInfo = mediaId?.let { StarrySky.get().mediaQueueProvider.getSongInfo(it) }
        doCall(songInfo)
    }

    fun resetValidIndex() {
        validIndex = 0
    }

    fun clear() {
        validIndex = 0
        mValidQueue.clear()
    }
}